### PR TITLE
build a CIRCUITPY_ERASER variant

### PR DESCRIPTION
**Pull request by dhalbert**
For each board, this builds one special build that erases CIRCUITPY when loaded. The status LED starts at red and goes to blue when done. The build does nothing after erasing: no USB mounting, etc. So it's language-independent.

The GitHub Actions build will create a new `CIRCUITPY_ERASER` directory in S3 for each board, alongside the language directories.

Basically, this is a simple way of making per-board flash erasers without involving Arduino, etc.

Tested on a Trinket M0, but I need to see GitHub Actions working properly too.

### CODE REVIEW

File: supervisor/shared/filesystem.c [hunk 1]
```diff
@@ -71,7 +71,7 @@ static void make_empty_file(FATFS *fatfs, const char *path) {
 
 
 static void make_sample_code_file(FATFS *fatfs) {
-    #if CIRCUITPY_FULL_BUILD
+    #if 1
```
Comment (hierophect) [inline supervisor/shared/filesystem.c:74]: Did you want to keep this, or can the whole if block be removed?

**Review by hierophect — COMMENTED** [review]
Thanks for making this. One alternative that I've used for scenarios that needed this is to instruct Circuitpython to enter the repl *first*, before executing any user code. Simply erasing without having the user do anything is conveniently fire-and-forget, but it's another option to consider - sitting in the REPL would let you access the filesystem and extract things you want from it before erasing.

File: supervisor/shared/filesystem.c (hunk 1 above)
Comment (dhalbert) [reply supervisor/shared/filesystem.c:74]: oops, that was a testing thing. It's only 75 bytes to add a non-empty code.py, but that can be a separate PR.

**Review by hierophect — APPROVED** [review]
@dhalbert makes sense, sounds good to me

**Comment by dhalbert** [comment]
making it draft while testing builds, not sure it's right (but build is being made)

**Review by tannewt — CHANGES_REQUESTED** [review]
I *really* don't like this. It has nothing to do with CircuitPython. I'd much rather factor out the useful bits from CircuitPython so that they can be used for other cross-board software. I understand this is the more expedient option but I think the extra work put into splitting useful things out will be more useful than we think.

It also erases all of the filesystem when we just want to disable boot.py and code.py to recover from USB mis-settings. This is what safe-mode is for and if it's broken on some boards, then we should fix it.

**Comment by hierophect** [comment]
> It also erases all of the filesystem when we just want to disable boot.py and code.py to recover from USB mis-settings. This is what safe-mode is for and if it's broken on some boards, then we should fix it.

Is there a way to force entry to safe mode from the hardware? I've had to build versions of Circuitpython like this (or revert to old builds of Circuitpython that don't support what I'm working on) for peripheral-related bugs triggered by code.py that hang without crashing into safe mode, or occasionally for code that does something like restarting too fast, like a code.py that just calls `microcontroller.reset()` right away or something. So I think there's a use case. But I do get your argument that it isn't really Circuitpython.

**Comment by dglaude** [comment]
Not sure if it has to be in or out of CircuitPython, but I believe Adafruit should be providing a safe UF2 file to restore the board to a working state, or one to test the board outside of user error...
If done here, it will benefit all of the non Adafruit board too, and every new board will automatically benefit from it.

**Comment by Neradoc** [comment]
> It also erases all of the filesystem when we just want to disable boot.py and code.py to recover from USB mis-settings. This is what safe-mode is for and if it's broken on some boards, then we should fix it.

Boards without a status LED are an issue. Plus the Raspberry pico doesn't even have a reset button, so you couldn't do anything without adding wires. The only "safe" state the user can achieve with a naked pico is bootloader.
There is the nuke UF2 of course, but...

Is it possible for those cases to have a UF2 that just writes some configuration bits that Circuitpython can then read ? Like "Drop this UF2 to start in safe mode". And do it again to restore normal behavior.

**Comment by tannewt** [comment]
> > It also erases all of the filesystem when we just want to disable boot.py and code.py to recover from USB mis-settings. This is what safe-mode is for and if it's broken on some boards, then we should fix it.
> 
> Is there a way to force entry to safe mode from the hardware? I've had to build versions of Circuitpython like this (or revert to old builds of Circuitpython that don't support what I'm working on) for peripheral-related bugs triggered by code.py that hang without crashing into safe mode, or occasionally for code that does something like restarting too fast, like a code.py that just calls `microcontroller.reset()` right away or something. So I think there's a use case outside USB. But I do get your argument that it isn't really Circuitpython.

On most boards you can click reset during a 700ms window at startup to restart into safe mode. On the ESP32-S2 the BOOT pin might work that way instead.

> Boards without a status LED are an issue. Plus the Raspberry pico doesn't even have a reset button, so you couldn't do anything without adding wires.

I think it's very rare to not have any status LED and even without it, the safe mode click should work (when a button exists.) We should fix the Pico to use the BOOTSEL state.

**Comment by dhalbert** [comment]
I'd say we could use one more of the following:
1. CIRCUITPY eraser - the CircuitPython build system knows the board configurations so that's why I based this on CPy. A "board database" would be nice in the long run, but it's not holding up progress. This requires only one build per board because it is not translation-dependent.
2. A safe-mode-only build. This requires doubling the number of artifacts, which will significantly increase CI time. I would rather not do that.
3. A build that does nothing but rename `boot.py` and `code.py` so that they will not run. It does not need to go to the REPL or similar, so it can be language-independent, so one build per board.
4. Make getting into safe mode easier. Maybe write up some language about how to do a "slow double-click" using the current 700ms timeout, or lengthen the timeout slightly. For boards that don't have the safe-mode click, fix that. Any trickier way of getting into safe mode like grounding pins, etc. will not work for projects that are in enclosures, etc.

**Comment by dhalbert** [comment]
Closing for now.

**Comment by hierophect** [comment]
> I think we could patch builds instead of rebuilding to do this. (Have a variable to check for safe mode and use the map file to know where it is in the binary and then change it.) It'd still double the number of artifacts but not the build time.

I like this idea - if it works, it achieves all the objectives in this thread, without the CI drawback.

**Comment by rjsdotorg** [comment]
> I'd say we could use one more of the following:
> 
> 1. CIRCUITPY eraser - the CircuitPython build system knows the board configurations so that's why I based this on CPy. A "board database" would be nice in the long run, but it's not holding up progress. This requires only one build per board because it is not translation-dependent.

Is there a how-to for this yet? I think anything that can be installed should have an uninstaller
We have a Metro M0 board that we cannot get to load an Arduino sketch - ie, we want to _**remove**_ OEM CircuitPython.
We have tried the power-up and double-click etc on [learn.adafruit.com](learn.adafruit.com/welcome-to-circuitpython/uninstalling-circuitpython ) and simply have gotten various errors upon upload failure.
Same system loads same sketches on Adalogger, XBoard etc.

In short, how to nuke/remove CircuitPython and leave it in a state for Arduino IDE?

**Comment by deshipu** [comment]
@rjsdotorg this bug was about erasing the contents of the filesystem on the flash, not CircuitPython itself. You can erase CircuitPython easily enough by simply flashing any Arduino program.

If your Arduino IDE can't flash it, it's not CircuitPython's fault.

**Comment by dhalbert** [comment]
@rjsdotorg Please bring this up in https://forums.adafruit.com or in discord (https://adafru.it/discord) and we can help diagnose. There are many possible causes for this.

**Comment by rjsdotorg** [comment]
> @rjsdotorg Please bring this up in https://forums.adafruit.com or in discord (https://adafru.it/discord) and we can help diagnose. There are many possible causes for this.

Will do. Thanks.
